src/components/MapCanvas.vue

```diff
@@ -18,49 +27,64 @@ import Konva from 'konva'
+        this.$data.tools = this.$store.getters.tools
+        const found = this.$data.tools.find((tool: ITool) => tool.enabled)
+        if (found) {
+          this.$data.showPing = false
+          this.$data.activeTool = found.name
+        } else {
+          this.$data.activeTool = ''
+        }
+      }
+    })
+  },
   methods: {
-    onClickHandler (e:any) : void {
-      // this.$store.dispatch('setCursorPosition', {
-      //   x: e.evt.x,
-      //   y: e.evt.y
-      // })
-      this.$data.tools = this.$store.getters.tools
-      const found = this.$data.tools.find((tool: ITool) => tool.enabled === true)
-      if (found) {
-        const stage = this.$refs.stage.getStage()
-        const layer = this.$refs.layer.getNode()
-        switch (found.name) {
-          case 'ping':
-            const id = Math.random()
-            const amplitude = 25
-            const period = 500
-            let radius = 0
-
-            let item = new Konva.Circle({
-              x: e.evt.x,
-              y: e.evt.y,
-              radius: radius,
-              stroke: 'red',
-              fill: 'white',
-              strokeWidth: 5
-            })
-
-            const anim = new Konva.Animation((frame: any) => {
-              radius = amplitude * Math.sin((frame.time * Math.PI) / 1000)
-              layer.add(item)
-            }, layer)
-            anim.start()
-            setTimeout(() => {
-              console.log('stopped')
-              radius = 0
-              anim.stop()
-            }, period)
-            // stage.draw()
+    // UNCOMMENT AND PLEASE STOP THIS FROM COMPLAINING
+    /*
+      Property 'throttledMethod' is incompatible with index signature.
+      Type '((this: never, e: any) => void) & Cancelable' is not assignable to type '(this: Vue, ...args: any[]) => any'.
+      The 'this' types of each signature are incompatible.
+        Type 'Vue' is not assignable to type 'never'.
+    */
+    // throttledMethod: _.throttle(function (e) {
+    //   this.onMouseMoveHandler(e)
+    // }, 75),
+    onMouseUpHandler (e: any) : void {
+      // @ts-ignore - property 'getNode' does not exist on type 'Vue | Element | Vue[] | Element[]'
+      const layer = this.$refs.layer.getNode()
+      switch (this.$data.activeTool) {
+        case 'ping':
+          this.$data.showPing = false
+          break
+      }
+    },
+    onMouseDownHandler (e: any) : void {
+      // @ts-ignore - property 'getNode' does not exist on type 'Vue | Element | Vue[] | Element[]'
+      const layer = this.$refs.layer.getNode()
+      switch (this.$data.activeTool) {
+        case 'ping':
+          this.$data.showPing = true
+          addPing(e, layer)
+          break
+      }
+    },
+    onMouseMoveHandler (e:any) : void {
+      // @ts-ignore - property 'getNode' does not exist on type 'Vue | Element | Vue[] | Element[]'
+      const layer = this.$refs.layer.getNode()
+      switch (this.$data.activeTool) {
```

Comment:
Best to think about a way to avoid duplicating the switch statement in each event handler for the active tool.

We could use the strategy pattern for the active tool. That way all you have to do is call `this.$data.activeTool.mouseMoveHandler(e, layer);`

Reply:
Not necessarily something we'd do now, but something we should do not too far in the future

Reply:
I will have a look and add something in, easy spot

Reply:
Managed to do it, was quite easy in the end. Now just to sort out all the bloody typescript errors

```suggestion
    onMouseDownHandler (e: any) : void {
      // @ts-ignore - property 'getNode' does not exist on type 'Vue | Element | Vue[] | Element[]'
      const layer = this.$refs.layer.getNode()
      this.$data.enableTool = true
      const activeTool = this.$data.activeTool
      // @ts-ignore
      this[activeTool]().onClick(e, layer)
    },
    onMouseMoveHandler (e:any) : void {
      // @ts-ignore - property 'getNode' does not exist on type 'Vue | Element | Vue[] | Element[]'
      const layer = this.$refs.layer.getNode()
      const activeTool = this.$data.activeTool
      if (this.$data.enableTool) {
        // @ts-ignore
        this[activeTool]().onClick(e, layer)
      }
    }
```